File: src/api/categories.ts
import apiClient from './client'
import type { Category, CreateCategoryPayload, UpdateCategoryPayload } from '../types'

export const categoriesApi = {
  list: async (): Promise<Category[]> => {
    return (await apiClient.get<Category[]>('/categories')).data
  },

  create: async (payload: CreateCategoryPayload): Promise<Category> => {
    return (await apiClient.post<Category>('/categories', payload)).data
  },

  update: async (id: number, payload: UpdateCategoryPayload): Promise<Category> => {
    return (await apiClient.patch<Category>(`/categories/${id}`, payload)).data
  },

  delete: async (id: number): Promise<void> => {
    await apiClient.delete(`/categories/${id}`)
  },
}

File: backend/app/routers/categories.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.database import get_db
from app.models import Category, Project
from app.schemas import CategoryCreate, CategoryUpdate, CategoryResponse

router = APIRouter(prefix="/categories", tags=["categories"])


@router.get("", response_model=list[CategoryResponse])
def get_categories(db: Session = Depends(get_db)):
    return db.query(Category).order_by(Category.sort_order).all()


@router.post("", response_model=CategoryResponse, status_code=201)
def create_category(data: CategoryCreate, db: Session = Depends(get_db)):
    max_order = db.query(Category).order_by(Category.sort_order.desc()).first()
    new_order = (max_order.sort_order + 1) if max_order else 0
    
    category = Category(name=data.name, sort_order=new_order)
    db.add(category)
    db.commit()
    db.refresh(category)
    return category


@router.patch("/{category_id}", response_model=CategoryResponse)
def update_category(category_id: int, data: CategoryUpdate, db: Session = Depends(get_db)):
    category = db.query(Category).filter(Category.id == category_id).first()
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    
    if data.name is not None:
        category.name = data.name
    if data.sort_order is not None:
        category.sort_order = data.sort_order
    
    db.commit()
    db.refresh(category)
    return category


@router.delete("/{category_id}", status_code=204)
def delete_category(category_id: int, db: Session = Depends(get_db)):
    category = db.query(Category).filter(Category.id == category_id).first()
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    
    db.query(Project).filter(Project.category_id == category_id).update({"category_id": None})
    db.delete(category)
    db.commit()
    return None
